src/util/simplify_expr_int.cpp

```diff
@@ -1248,6 +1248,15 @@ bool simplify_exprt::simplify_inequality(exprt &expr)
         return false;
       }
     }
+    else if(tmp0.type().id() == ID_string)
+    {
+      const auto tmp0_const = expr_checked_cast<constant_exprt>(tmp0);
+      const auto tmp1_const = expr_checked_cast<constant_exprt>(tmp1);
+
+      bool equal = (tmp0_const.get_value() == tmp1_const.get_value());
+      expr.make_bool(expr.id() == ID_equal ? equal : !equal);
+      return false;
+    }
```

Comment:
I think this is somewhat questionable now: This code may be called with expr.id() == {ID_ge, ID_lt, ...} and not just ID_equal/ID_notequal. Hence I think the guard for this branch should not be on `type.id() == ID_string`, but instead be on the operator, irrespective of the type. And then this is very inviting to refactor the calls  to `expr_checked_cast` as really all the branches around would benefit from that step. Finally you could then get rid of all the ID_equal/ID_notequal cases in the other branches around here.

Reply:
True. Yet another argument for type checking our expressions to make hidden assumptions explicit. Some operators cannot be used with certain types; such expressions should never be generated.